gCode Legend & Viewer Improvements

yw4z opened the pull request:
### CHANGES
**For decreasing window width and increasing visible screen space for model**
• Header > Used % instead "Percent" and removed "%" from types
• Header > Used "Usage" instead "Filament usage". Requires translation
• Removed spaces between [Length][Unit] and [Weight][Unit]
• Used eye icon and removed "Display" from header
• Removed "color scheme" text to compact frame width. Writing on different line can be an alternative method instead hiding

**QOL improvements**
• Added button for toggling gCode editor. Fixes https://github.com/SoftFever/OrcaSlicer/issues/3136
• Increased Combo box's dropdown menu height to prevent scrolling to reach last elements of list. Fixes https://github.com/SoftFever/OrcaSlicer/issues/4486 https://github.com/SoftFever/OrcaSlicer/issues/6359
• Stored window width to keep Expand/Collapse button on same position to prevent mouse movement requirement while toggling

**Visual Improvements**
• Removed background color for expand/collapse button
• Added new icons for Expand/Collapse/gCode Viewer Toggle
• Removed darker rectangle on top. That gave more clean look imo
• Added window radius to match style a bit with gizmos. That gave more modern look imo
• Added a small gap between main toolbar and legend
• Added a small gap between legend and gcode viewer
• Fixes some of problems that related scaling

### TODO
• Add scroll ability to gcode editor

### COMPARISON
before
![Screenshot-20250126172934](https://github.com/user-attachments/assets/9f04d98e-ec37-4645-9916-9083ea1629a4)
After -  Width of legend is smaller %15 percent and allows for more screen space
![Screenshot-20250220205034](https://github.com/user-attachments/assets/93a6bcd5-8447-4e5f-b645-430821cdce70)

before
![Screenshot-20250126233848](https://github.com/user-attachments/assets/1d1428f8-5c66-4b07-97b8-86b24f738ed5)
after - Dropdown height increased to fit all contents. This prevents unnecessary scrolling
![Screenshot-20250220204950](https://github.com/user-attachments/assets/8876a22b-62aa-43fe-8bb3-70a519c8a357)

before
![Screenshot-20250127004637](https://github.com/user-attachments/assets/5c68989f-fcdc-4316-a6c8-eb66042bc578)
after
![Screenshot-20250220204845](https://github.com/user-attachments/assets/988db7d4-6e25-4418-b70f-6650889f4eab)

before - expand button changes position that requires unnecessary mouse movements
![orca-slicer_VsZsIkYHZx](https://github.com/user-attachments/assets/60acae96-6475-4c21-9eda-645eb4c52099)
after - expand/fold button stays at same position
![orca-slicer_9kgTF40qeK](https://github.com/user-attachments/assets/3293c5aa-3cb7-407f-b937-d31cb331b4d3)

before
![Screenshot-20250220204243](https://github.com/user-attachments/assets/1590a8cf-ff6d-4bd5-8f27-cb864e513ac4)
after - added gaps between windows that shows UI more organized and modern 
![Screenshot-20250220204254](https://github.com/user-attachments/assets/f345dfc0-f640-4cdc-8975-85f0423061ca)

Side by side
![Screenshot-20250220213854](https://github.com/user-attachments/assets/0d637c32-6274-436f-bb88-daeff3196579)

buzzhuzz commented:
My question is not directly related to this PR, but have you ever considered making G-code inspector colors similar to the main widgets themes? Somewhat similar to the gizmo tool bar.

yw4z commented:
thanks for indicating that point. i will look what i can do

discip commented:
@yw4z
First of all:
Thank you for taking on this work! 😊
Most of the changes you make seem to make sense while some are not so understandable.

For example, why did you decide to replace the checkbox with an eye icon whose state can only be determined by the color, while the checkbox lets you know immediately if it has been checked or not, even if you cannot distinguish between colors?

The main goal of every single change should be to improve usability.

I once read a quote in another repo that went something like this:
**_`"good UI is invisible"`_**

However, this particular change (eye icon) makes it a bit harder to use for some than before.

In this case, please either keep the checkboxes, or make the difference between the "selected" and "unselected" states more obvious, e.g. by representing the unchecked state by a closed eye.
_I prefer the checkboxes however._

kind regards

kekkodance commented:
this looks nice.

yw4z commented:
hi @discip
i'm not trying to remove everything. Just removing unnecessary things to get more cleaner/modern look. 
I assume "Less is more" is better fits if you find a quote.
Main goal is reducing area of legend since not everyone uses 24+inch monitor.
gCode viewer toggle button also for that so people can easily hide it to get more space when it needed. Especially placed it near to expand/collapse button since their purpose is similar

here is an alternative. used boxes on non-visible items. 
![Screenshot-20250127024449](https://github.com/user-attachments/assets/6490f486-bad1-4a48-8ae9-2311c449e2f8)

as you know eye icon already represents visibility on most UI's. Let's hear what people think

buzzhuzz commented:
@yw4z , could you please try crossed-out gray eye for disabled items?

discip commented:
@yw4z
> i'm not trying to remove everything. Just removing unnecessary things to get more cleaner/modern look.

This is not what I wrote. 😊
The only thing I addressed was the **eye icon**.

> Main goal is reducing area of legend since not everyone uses 24+inch monitor.

Yes, on that one we are on the same page. And that's one of the things I thanked you for.

> here is an alternative. used boxes on non-visible items.

Mixing these themes doesn't look right, as you would expect either an _empty box_ & _one with a check mark in it_, or an _opened eye_ and a _closed one_. These are completely different concepts.

yw4z commented:
@buzzhuzz 
![Screenshot-20250127004600 Copy](https://github.com/user-attachments/assets/78d9c17a-4566-45b3-b2a2-1437b3434e6b)

@discip here is example from adobe. Using boxes not completely unrelated because guides users to a clickable area or shows them as unchecked
![Screenshot-20250127132703](https://github.com/user-attachments/assets/ee67db1b-1bb6-4da3-952c-5127933d2ecc)

igiannakas commented:
haven't tested yet but does it adapt for a larger width when multi material prints are displayed?

yw4z commented:
@igiannakas didnt saw issues when using 2 material but didnt tested further. Window width determined by bottom content and dynamic. i assume it shouldnt be a problem

igiannakas commented:
Awesome! 

The "eye" / "not eye" I think is more consistent. It conveys the meaning of "visible" and not visible when it is greyed out. Mixing tick box and the eye icon is less continuous from a user experience. We shouldn't be changing the type of box I feel.

Hence, I vote for this one:

![image](https://github.com/user-attachments/assets/028452da-2a01-420f-9cb9-850dfbdae80f)

yw4z commented:
@igiannakas i didnt touched these spacings because it broke many placement when i changed before

yw4z commented:
@buzzhuzz btw which os you use? eye icon looks blurry on your image on https://github.com/SoftFever/OrcaSlicer/pull/8294#issuecomment-2629638636
or is that caused by screenshot quality? i draw icons with code instead svg icons. maybe i have to switch to vector icons

SoftFever commented:
@yw4z 
These are awesome improvements.
Thank you!

yw4z commented:
Updated visuals on first post with latest changes and added more comparisons

@SoftFever 
i assume thats all i can improve right now. reduced code changes as much as possible and added detailed comments for each change

Only i couldnt figure out how to add scrolling support to gcode viewer section. that could be a nice addition. it's same thing using slider on bottom but with scrolling on viewer

Also this commit https://github.com/SoftFever/OrcaSlicer/pull/8294 will give another fix for longer translations of "Usage". you can find more information on that commit

I marked commit as ready. let me know your thoughts

GlauTechCo commented:
Hello, macbook m1 2020  13.3 inc

Default screen resolution( 1440x900 )
<img width="381" alt="Ekran Resmi 2025-02-20 23 57 14" src="https://github.com/user-attachments/assets/4ecafb6d-01cb-45be-be7d-8e57e9716a8f" />


Screen resolution( 1680x1650 )
<img width="449" alt="Ekran Resmi 2025-02-21 00 05 02" src="https://github.com/user-attachments/assets/683fea54-d517-4dc2-90e8-6d9ee8cd5a04" />


Screen resolution( 2560x1600 )
![Ekran Resmi 2025-02-21 00 05 19](https://github.com/user-attachments/assets/d7722c7c-7a07-4fe8-ba89-18cc6dc5d64b)

yw4z commented:
@GlauTechCo can you test with latest commit. corrected scaling for top buttons and visible/hidden icons. there are many scaling related issues exist probably i will try to fix them in future

GlauTechCo commented:
> @GlauTechCo can you test with latest commit. corrected scaling for top buttons and visible/hidden icons. there are many scaling related issues exist probably i will try to fix them in future

Thnx
<img width="429" alt="Ekran Resmi 2025-02-21 09 36 40" src="https://github.com/user-attachments/assets/88c39c38-94b8-4b1d-8e1c-fcd46370103a" />

afmenez commented:
I liked all the changes, except for the removal of the '%' on the 3rd column.  The size gain is minimal, but the data is less clear without it.  Note that all other columns have a unit too.

yw4z commented:
removing spaces between units(m/g) and numbers was minimal change too but applying all together saves good amount space

```
1. 4m32s  32.4  0.60  1.79  eye               
2. 4m32s  32.4  0.60m  1.79g  eye         
3. 4m32s  32.4%  0.60m  1.79g  eye 
4. 4m32s  32.4%  0.60 m  1.79 g  eye 
```
1. also tried removing m/g but it makes unreadable
2. current
3. added % symbol. this symbol has full width letter size. wider than any letter
4. without any changes

header already says its % why want to see that in content?

afmenez commented:
I don't want to look up on the column name to understand what is the data.  You could also argue that the 'm' and 'g' are redundant, but it is quicker to understand when they are there on each line.
Thinking of it now, the column could have an abbreviation like "Perc." instead of just "%".

TomPcz commented:
> • Removed spaces between [Length][Unit] and [Weight][Unit]

Please revert this change. Numerical values and unit symbols should be separated by a space (see ISO 31-0). Also, could you please correct the same issue in the time column and the total time?

yw4z commented:
@afmenez @TomPcz i recommend you both to build and use it for while

TomPcz commented:
I have tried. Honestly, I have no need to save space on my screen :)
![Screenshot](https://github.com/user-attachments/assets/8f3581fb-2425-41f0-b538-ba0dc33190e4)

I understand that users with smaller screens might have an issue, but they can always hide the legend. That shouldn't be an excuse for ignoring the ISO standard. As a physicist, I can say that not having space between the number and the unit just feels wrong and unnatural.

Moreover, readability is far more important than saving a few pixels. Those numbers are crucial when optimizing a printing profile (for speed/weight). In my opinion, having all those values so close together makes them harder to read.

I have found an issue with the "eye" icon. It does not scale properly when I use screen scaling (Linux openSUSE Tumbleweed, KDE, Wayland). This is at 120 % (here only the left part of the icon is clickable):
![Snímek obrazovky_20250227_181840](https://github.com/user-attachments/assets/f3632ba1-cdb0-45be-8e80-04c96c8a3f62)
And at 150 % (here even the show/hide gcode icon is cropped):
![Snímek obrazovky_20250227_182638](https://github.com/user-attachments/assets/74538545-6d8b-4809-a819-378f22ffa1be)

discip commented:
> 3. 4m32s  32.4%  0.60m  1.79g  eye

The 3rd Option is best!
No need for spaces between value and units. In this case, the screen-space is more valuable than compliance with standards.
Perhaps you could use a different color for the units to improve readability.

Aligning the various values ​​on the right would also help:

![image](https://github.com/user-attachments/assets/81c0b0de-01a7-442d-a1c4-dd09ca568dac)

discip commented:
> Almost perfect, I would add a leading zero on "5m08s".

Since there is no decimal separator the leading **`0`** can be omitted.

more space = better readability 😊

yw4z commented:
sent new changes for
• fixing scaling related issues
• increased spacing between columns
• reduced required code changes for hiding "Display" header

![Screenshot-20250228024141](https://github.com/user-attachments/assets/1347959f-49f6-4cbe-be7c-053244d6d72a)

increased window width a bit but readability increased. its a trade off but window width still less then original code
make a test and share your thoughts

Weeris102 commented:
can you add a space between minutes and seconds?
"23m 46s" instead of "23m46s"

LightDestory commented:
I started using this PR-based Orca for a while now, I got used to the new view pretty fast. I had to run the beta2 for a vertical support and as soon as I looked on the old view I was shocked. This PR is a must for me now :D


Thanks @yw4z !

I pray for this PR to get merged soon @SoftFever

SoftFever commented:
Very nice touch, thank you so much!

rubienr commented:
Nice enhancement.

I face an issue which might be related to 9fe905c47. After slicing is finished and the legend view is rendered, the view starts to expand horizontally (with my settings from right to left) until it reaches the left widow border and a little bit beyond.

![image](https://github.com/user-attachments/assets/bf48244d-9df0-49b9-8596-e3d94aaffc4d)

My hotfix is to click the collapse button (hard to catch while moving). Once minimized, the horizontal expansion stops.

Steps to reproduce:
1. open new project
2. load a simple model
3. click "Preview" -> slicing starts
4. when slicing is finished, the legend view is rendered which starts expanding beyond window border

```bash
lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 24.10
Release:	24.10
Codename:	oracular
```

ancient X11 (not Wayland)

```bash
xrandr
Screen 0: minimum 8 x 8, current 5120 x 2160, maximum 32767 x 32767
HDMI-0 disconnected (normal left inverted right x axis y axis)
DP-0 connected primary 5120x2160+0+0 (normal left inverted right x axis y axis) 64mm x 27mm
   2560x1080     60.00 +
[...]
```

discip commented:
Almost the same here on Manjaro Gnome.
In my case, the legend only expands while the mouse is moved.

discip commented:
@yw4z
@SoftFever
Almost the same here on Manjaro Gnome.
In my case, the legend only expands while the mouse is moved.

dewi-ny-je commented:
All nice but "</>" is as good as dropping it altogether.
Honestly how many people would understand what it means? I am far from being a coder and far from bein a newbie in computer stuff, and I cannot understand the meaning.

Using "Colour by:" is not long and it would be understood, or drop it.
But not "</>", please.

igiannakas commented:
Quick feedback - 

![image](https://github.com/user-attachments/assets/1048bbb1-3b96-4e96-8ab9-d73e7d22d6cd)

Enabling/ disabling the visibility of a feature necessitates clicking on the line of that feature - it doesnt work when clicking on the new "eye" button. Not the end of the world, but was counter intuitive when I first tried it, to the extent that I thought the feature wasn't working properly :D

yw4z commented:
i checked expansion issue on windows but not exist, something should be different in linux while rendering text or icon that effects width. i will recheck code for better solution. i saw this issue while editing code. it happens when content doesnt fit. it automatically expands width endlessly

@dewi-ny-je  about icon
same icon used on right slider & parameters boxes so its not unique to this place. it's ok to ppl get use to over time

yw4z commented:
@igiannakas this happens on existing code too. couldnt find what blocking mouse click. it works on top 5 rows without problems but sometimes not working bottom rows. if you slice second times they work again.

yw4z commented:
@rubienr @discip does window expands only while scrollbar exist? or even without scroll bar

rubienr commented:
> @rubienr @discip does window expands only while scrollbar exist? or even without scroll bar

It also happens without vertical legend-view scrollbar but only for:
- Line Type
- Filament
- Speed

It work as expected for:
- Layer Height
- Line Width 
- Flow
- Layer Time
- Layer Time (log)
- Fan Speed 
- Temperature

![image](https://github.com/user-attachments/assets/c01eab91-7dfd-468e-aaef-d35777d2752c)

rubienr commented:
> Can you build from source? i dont have linux based setup so need help from other developers [...]

Yes.

With your patch in `GCodeViewer.cpp:4475` and `GCodeViewer.cpp:4526` it hehaves as expected (not expanding any more). 
Shall i also test with either `4475` or `4526` patched?

Anther thing i noticed yet: the toggle gcode-viewer icon is slightly clipped on the right side:
![image](https://github.com/user-attachments/assets/8afb351e-ba97-4e07-a3c6-9cf3f9f92cba)

rubienr commented:
This chunk fixes the issue:
```patch
diff --git a/src/slic3r/GUI/GCodeViewer.cpp b/src/slic3r/GUI/GCodeViewer.cpp
index dad2dfb36..965e930de 100644
--- a/src/slic3r/GUI/GCodeViewer.cpp
+++ b/src/slic3r/GUI/GCodeViewer.cpp
@@ -4472,7 +4472,7 @@ void GCodeViewer::render_legend(float &legend_height, int canvas_width, int canv
                 //ImGui::Checkbox(("##" + columns_offsets[0].first).c_str(), &visible);
                 //ImGui::PopStyleVar(1);
                 // ORCA replace checkboxes with eye icon
-                ImGui::SameLine(ImGui::GetWindowWidth() - (16.f + 0.f) * m_scale - window_padding * 2 - (ImGui::GetScrollMaxY() > 0.0f ? ImGui::GetStyle().ScrollbarSize : 0));
+                ImGui::SameLine(ImGui::GetWindowWidth() - (16.f + 6.f) * m_scale - window_padding * 2 - (ImGui::GetScrollMaxY() > 0.0f ? ImGui::GetStyle().ScrollbarSize : 0));
                 ImGui::Text(into_u8(visible ? ImGui::VisibleIcon : ImGui::HiddenIcon).c_str(), ImVec2(16 * m_scale, 16 * m_scale));
             }
         }
```

Just this chunk does not fix the issue:
```patch
diff --git a/src/slic3r/GUI/GCodeViewer.cpp b/src/slic3r/GUI/GCodeViewer.cpp
index dad2dfb36..9ea54da4a 100644
--- a/src/slic3r/GUI/GCodeViewer.cpp
+++ b/src/slic3r/GUI/GCodeViewer.cpp
@@ -4523,7 +4523,7 @@ void GCodeViewer::render_legend(float &legend_height, int canvas_width, int canv
         for (size_t i = 0; i < title_offsets.size(); i++) {
             if (title_offsets[i].first == _u8L("Display")) { // ORCA Hide Display header
                 ImGui::SameLine(title_offsets[i].second);
-                ImGui::Dummy({(16.f - 6.f) * m_scale, 1}); // 16(icon) - 6(half of spacing)
+                ImGui::Dummy({16.f * m_scale, 1});
                 continue;
             }
             ImGui::SameLine(title_offsets[i].second);
```

yw4z commented:
you can send that as commit

i will recheck for a better solution for the clipping issue on icons 
i assume you have scaling or hiDPI monitor that requires scaling

yw4z commented:
@rubienr ci assume i found problem for clipping icon
GCodeViewer.cpp line 4661 : can you replace and test
` float calc_padding = std::max((ImGui::GetFrameHeight() - 16 * m_scale) / 2, 0.f);`

rubienr commented:
> you can send that as pull request

OK: https://github.com/SoftFever/OrcaSlicer/pull/9171

yw4z commented:
@rubienr did you test https://github.com/SoftFever/OrcaSlicer/pull/8198#issuecomment-2767525112

rubienr commented:
> @rubienr did you test [#8198 (comment)](https://github.com/SoftFever/OrcaSlicer/pull/8198#issuecomment-2767525112)

It still clipps a bit:
![image](https://github.com/user-attachments/assets/0da7641a-d341-4f23-a62b-65358cdbc0d8)

yw4z commented:
@rubienr can you send a screenshot while hovering button? i can understand better button size and paddings

rubienr commented:
@yw4z 
![image](https://github.com/user-attachments/assets/4d5fa9b7-e011-4b08-a52a-688ad671190c)
![image](https://github.com/user-attachments/assets/ba7dac85-ec72-49ed-89ca-b0c31f6d8821)

yw4z commented:
Can you test this. this values exact same as combo box
GCodeViewer.cpp line 4661
` float calc_padding = std::max(((ImGui::GetStyle().FramePadding.y * 2 + ImGui::CalcTextSize("X").y) - 16 * m_scale) / 2, 0.f);`

rubienr commented:
Same result: 

![image](https://github.com/user-attachments/assets/da3a4b1a-4bd5-4980-9f9f-9b03017fe53f)
![image](https://github.com/user-attachments/assets/be89dae5-eb5a-48e7-bb08-961735574041)

Chagne:
```patch
diff --git a/src/slic3r/GUI/GCodeViewer.cpp b/src/slic3r/GUI/GCodeViewer.cpp
index 965e930de..48b67cfcb 100644
--- a/src/slic3r/GUI/GCodeViewer.cpp
+++ b/src/slic3r/GUI/GCodeViewer.cpp
@@ -4658,7 +4658,7 @@ void GCodeViewer::render_legend(float &legend_height, int canvas_width, int canv
     ImGui::PushStyleColor(ImGuiCol_Button, ImVec4(0, 0, 0, 0));
     ImGui::PushStyleColor(ImGuiCol_ButtonHovered, ImVec4(84 / 255.f, 84 / 255.f, 90 / 255.f, 1.f));
     ImGui::PushStyleColor(ImGuiCol_ButtonActive, ImVec4(84 / 255.f, 84 / 255.f, 90 / 255.f, 1.f));
-    float calc_padding = (ImGui::GetFrameHeight() - 16 * m_scale) / 2;                      // ORCA calculated padding for 16x16 icon
+    float calc_padding = std::max(((ImGui::GetStyle().FramePadding.y * 2 + ImGui::CalcTextSize("X").y) - 16 * m_scale) / 2, 0.f); // ORCA calculated padding for 16x16 icon
     ImGui::PushStyleVar(ImGuiStyleVar_FramePadding, ImVec2(calc_padding, calc_padding));    // ORCA Center icon with frame padding
     ImGui::PushStyleVar(ImGuiStyleVar_FrameRounding, 3.0f * m_scale);                       // ORCA Match button style with combo box
```
